plume-models/src/posts.rs

```diff
@@ -925,154 +873,365 @@ impl Post {
     type Error = Error;
+    type Object = LicensedArticle;
 
-    fn from_activity(
-        (conn, searcher): &(&'a Connection, &'a Searcher),
-        article: LicensedArticle,
-        _actor: Id,
-    ) -> Result<Post> {
+    fn from_db(c: &PlumeRocket, id: &str) -> Result<Self> {
+        Self::find_by_ap_url(&c.conn, id)
+    }
+
+    fn from_activity(c: &PlumeRocket, article: LicensedArticle) -> Result<Self> {
+        let conn = &*c.conn;
+        let searcher = &c.searcher;
         let license = article.custom_props.license_string().unwrap_or_default();
         let article = article.object;
-        if let Ok(post) =
-            Post::find_by_ap_url(conn, &article.object_props.id_string().unwrap_or_default())
-        {
-            Ok(post)
-        } else {
-            let (blog, authors) = article
-                .object_props
-                .attributed_to_link_vec::<Id>()?
-                .into_iter()
-                .fold((None, vec![]), |(blog, mut authors), link| {
-                    let url: String = link.into();
-                    match User::from_url(conn, &url) {
-                        Ok(u) => {
-                            authors.push(u);
-                            (blog, authors)
-                        }
-                        Err(_) => (blog.or_else(|| Blog::from_url(conn, &url).ok()), authors),
+
+        let (blog, authors) = article
+            .object_props
+            .attributed_to_link_vec::<Id>()?
+            .into_iter()
+            .fold((None, vec![]), |(blog, mut authors), link| {
+                let url: String = link.into();
+                match User::from_id(&c, &url, None) {
+                    Ok(u) => {
+                        authors.push(u);
+                        (blog, authors)
                     }
-                });
+                    Err(_) => (blog.or_else(|| Blog::from_id(&c, &url, None).ok()), authors),
+                }
+            });
 
-            let cover = article
-                .object_props
-                .icon_object::<Image>()
-                .ok()
-                .and_then(|img| Media::from_activity(conn, &img).ok().map(|m| m.id));
+        let cover = article
+            .object_props
+            .icon_object::<Image>()
+            .ok()
+            .and_then(|img| Media::from_activity(&c, &img).ok().map(|m| m.id));
 
-            let title = article.object_props.name_string()?;
-            let post = Post::insert(
+        let title = article.object_props.name_string()?;
+        let post = Post::insert(
+            conn,
+            NewPost {
+                blog_id: blog?.id,
+                slug: title.to_kebab_case(),
+                title,
+                content: SafeString::new(&article.object_props.content_string()?),
+                published: true,
+                license,
+                // FIXME: This is wrong: with this logic, we may use the display URL as the AP ID. We need two different fields
```

Comment:
is it related to #296 ?

Reply:
More or less I think. Adding a `url` field to posts could solve both issues.